proposals/route-blinding.md

```diff
@@ -0,0 +1,308 @@
+
+Some use-cases where route blinding is useful include:
+
+* Recipient anonymity when receiving payments
+* Using unannounced channels in invoices without revealing them
+* Forcing a payment to go through a specific set of intermediaries that can witness the payment
+* Providing anonymous reply paths for onion messages
+
+### Overview
+
+At a high level, route blinding works by having the recipient choose an _introduction point_ and a
+route to himself from that introduction point. The recipient then blinds each node and channel
+along that route with ECDH. The recipient includes the blinded route and a _hop-blinding secret_ in
+the invoice, which allows each node in the blinded route to incrementally unblind the payloads.
+
+This scheme requires all the nodes in the blinded route and the sender to activate support for the
+feature. It only becomes effective once a big enough share of the network supports it.
+
+### Notations
+
+* A node `N(i)`'s `node_id` is defined as: `P(i) = k(i) * G` (`k(i)` is the node's private key).
+* Blinded `node_id`s are defined as: `B(i) = b(i) * G` (`b(i)` is the blinding factor).
+* Ephemeral public keys are defined as: `E(i) = e(i) * G`.
+
+### Requirements
+
+A node `N(r)` wants to provide a blinded route `N(0) -> N(1) -> ... -> N(r)` that must be used
+to receive onion messages.
+
+* The channels used along that route may be either announced or unannounced.
+* When used for payments, intermediate nodes in the blinded route MUST NOT learn `payment_secret`.
+* Intermediate nodes in the blinded route MUST NOT learn the `node_id` or `scid` of other
+  intermediate nodes except for their immediate predecessor or successor.
+* Intermediate nodes in the blinded route MUST NOT learn their distance to the recipient `N(r)`.
+* Senders MUST NOT learn the real `node_id` and `scid` of the blinded intermediate hops after the
+  introduction point `N(0)`.
+* If `N(r)` creates multiple blinded routes to herself, senders MUST NOT be able to tell that these
+  routes lead to the same recipient (unless of course this information is leaked by higher layers
+  of the protocol, such as using the same `payment_hash`).
+
+### Encrypted data
+
+Route blinding introduces a new TLV field to the onion `tlv_payload`: the `encrypted_data`.
+
+This field is used to carry data coming from the builder of the route that cannot be modified by the
+sender. For route blinding it only needs to contain the `scid` to use when forwarding the message,
+but it may be extended with additional data in the future. It uses ChaCha20-Poly1305 as AEAD scheme.
+
+1. type: 10 (`encrypted_data`)
+2. data:
+    * [`...*byte`:`encrypted_data`]
+
+Once decrypted, the content of this encrypted payload is a TLV stream that contains information to
+identify the next node.
+
+### Creating a blinded route
+
+`N(r)` performs the following steps to create a blinded route:
+
+```text
+Initialization:
+
+  e(0) <- {0;1}^256
```

Comment:
Could this be the shared secret of the introduction node from the regular onion that contains the blinded path?

Reply:
It can in some specific cases, but in general no, because this part is generated by the **recipient**, before the sender actually chooses a route and derives the shared secrets for the regular onion (so there would be a chicken and egg problem).

However, in the specific case of onion messages, when a **sender** uses route blinding _in the forward direction_ they could probably use the onion shared secret, but it deserves more thoughts. I'd rather keep them completely separate, that would ensure there's no "cross-layer" attack vector here.